### DISCUSSION TITLE
Localization support

### DISCUSSION THREAD

**Discussion by feikname** [piqueserver/piqueserver#633] [Feature Requests]
Add localization support to default scripts.

**Comment by kmsiapps** [comment]
Which part are you planning to add translation support? Does it support unicode chars, such as Korean Alphabet?

**Comment by feikname** [comment]
@kmsiapps The default scripts, which are visible to players in-game (like voteckick, motd). As long as the client support unicode chars (such as OpenSpades), it should work with unicode.

I am not sure how the original client would deal with it, but it'd probably print a lot of garbage.

Ideally, piqueserver should extend the 0.75 protocol in a backwards compatible-way to do some hand-shakes for checking this feature (and a few more). If we do take this route, we'll have to discuss it with yvt first.

**Comment by noway** [comment]
There is a script which detects whether the client is openspades client or not. And gets version of openspades etc.

That's like user agent checking, which is generally from soft eng perspective is not that good of a route. A possibility to check for specific features would be better (doesSupportUnicode packet for example)

**Comment by kmsiapps** [comment]
Great. But Openspades before version 0.1.0 doesn't have font-unicode.pak for default, so users have to manually put it to Resources folder to activate unicode support. So @noway421 's suggestion would be better for it.

**Comment by feikname** [comment]
@kmsiapps They're actully the same suggestion lol.

**Comment by kmsiapps** [comment]
@feikname I may have misunderstood your meaning ;) I just meant it's not enough to check if the client is OS or not, because unicode support should be done by user manually on OS.

BTW, is it available for piqueserver to support non-ASCII characters? Pyspades shows error like this : UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 0: ordinal not in range(128)

**Comment by feikname** [comment]
@kmsiapps In the terminal or in-game?

We're moving to Python 3, as so, full Unicode support is very likely to be implemented in the next month, even if there isn't one already.

**Comment by kmsiapps** [comment]
@feikname terminal. In-game shows nothing related to scripts because of the error.

**Comment by feikname** [comment]
Sorry, I didn't communicate well.

I mean, you typed unicode characters in the terminal or inside the game? If it was in-game, did you use OpenSpades?

Its quite likely we do have unicode support, because OpenSpades and PySnip shared some things in common. And we're basically a fork of PySnip.

PySpades is older software and I am almost sure it doesn't support unicode.

**Comment by kmsiapps** [comment]
@feikname Nope. Now I see my comment was weird because of my editing. Sorry ;_;
I edited spadenadefix.py to include non-ASCII chars, and the console shows that message.

**Comment by feikname** [comment]
@kmsiapps Judging by [this stackoverflow question](http://stackoverflow.com/questions/21129020/how-to-fix-unicodedecodeerror-ascii-codec-cant-decode-byte), the problem might be that you edited the file without the proper encoding. Try editing `spadenadefix.py` with Notepad++ and setting the encoding to *UTF-8*. 

If it doesn't work, then you should also try adding the `u` prefix to your string, like: 
```python
print(u"No glitches allowed!")
```

If that doesn't work either, then pyspades has no support for unicode, unfortunately.

**Comment by feikname** [comment]
If you provide the string you are trying to use, I can test it inside piqueserver in a couple of hours.

**Comment by kmsiapps** [comment]
@feikname I set it to UTF-8(Actually the server doesn't start witout it).
I also tried with pysnip(released 2013-11-27), and it shows the same result.

Also, when I typed non-ASCII chars on the console, chat says "??????" on Openspades.
Vise versa is the same, but when I type it on OS to global chat, another OS client can see it properly.

I'll try with up-to-date pysnip server (It has error like this, though)
> File "twisted\internet\base.pyo", line 1192, in run
          File "twisted\internet\base.pyo", line 1201, in mainLoop
          File "twisted\internet\base.pyo", line 824, in runUntilCurrent
          File "twisted\internet\task.pyo", line 218, in __call__
        --- <exception caught here> ---
          File "twisted\internet\defer.pyo", line 140, in maybeDeferred
          File "pyspades\server.pyo", line 1578, in update
          File "pyspades\server.pyo", line 1597, in update_network
          File "pyspades\contained.pyx", line 104, in pyspades.contained.WorldUpdate.items.__set__ (./pyspades/contained.cpp:5397)
        exceptions.TypeError: Expected dict, got list

**EDIT :** PySnip v0.75 Win32 Revision f8808c5 doesn't work on my computer. It says error above, and I can't connect to the server with localhost.

**Comment by kmsiapps** [comment]
@feikname I just edited ratio.py for test on Openspades, and used "ㅇㅇ".

**Comment by feikname** [comment]
@kmsiapps Thanks for the information!

**Comment by kmsiapps** [comment]
@feikname I tested again with u method(on disco.py) :
> If it doesn't work, then you should also try adding the u prefix to your string, like:
print(u"No glitches allowed!")

It works, but Openspades still says "??". Maybe Openspades' encoding method is different from piqueserver's?

**Comment by feikname** [comment]
You need to have Unicode enabled in OpenSpades besides having the pak, if I'm not mistaken.

**Comment by noway** [comment]
>BTW, is it available for piqueserver to support non-ASCII characters? Pyspades shows error like this : UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 0: ordinal not in range(128)

This is certainly bug. I remember pyspades was fine with showing the unicode in terminal. Maybe it was a custom patch, or we have broken something.

>I edited spadenadefix.py to include non-ASCII chars, and the console shows that message.

Can you please confirm that it happens only in that script if you manually write some international characters into string literals in source code? Trying to pin point the issue.

I remember i was looking into this: there had to be cp437 (extended ascii) encoding set instead of ascii. That's needed so that log shows correct encoding as does OpenSpades.

If you want to send something from a script, there should be the following written in the header of the python script:

```python
# -*- coding: utf-8 -*-
```
which is controlling comment for python which it uses to switch to unicode parsing. Here's a source: https://fedorahosted.org/releases/k/i/kitchen/docs/unicode-frustrations.html#frustration-3-inconsistent-treatment-of-output

**Comment by kmsiapps** [comment]
@noway421 
When I used self.send_chat("테스트") on the script the console shows :
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 0: ordinal not in range(128)

But if I use  self.send_chat(**u**"테스트"), it works with no error. But Openspades still says "???".
I set encoding as UTF-8 on Notepad++, and "Allow unicode" settings in Openspades are turned on.

**Comment by noway** [comment]
self.send_chat(u"테스트".encode('cp437'))

try that

**Comment by kmsiapps** [comment]
@noway421 Still doesn't work. It says : 
> UnicodeEncodeError: 'charmap' codec can't encode characters in position 0-2: character maps to undefined

**Comment by samuelallan72** [comment]
That's because it's unicode - `self.send_chat(u"테스트".encode('utf-8'))` should work.

**Comment by kmsiapps** [comment]
@swalladge Then it says :
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xed in position 0: ordinal not in range(128)

It's just the same as self.send_chat("테스트")

**Comment by feikname** [comment]
@kmsiapps  You're trying in which server, pyspades?

**Comment by kmsiapps** [comment]
@feikname I'm using pysnip-feature_server-8a845b0a4322 (2013.11.27) on bitbucket.

**Comment by noway** [comment]
lol i'd just try it all the possible ways, try `self.send_chat("테스트".encode('cp437'))` for example too

**Comment by kmsiapps** [comment]
@noway421 Still shows same error. In light of the fact that Openspades shows "???" when I typed "테스트" into console directly(and vise versa), i assume it's not just about the script encoding thing.

**Comment by kmsiapps** [comment]
When typed 테스트 on console :
![-1](https://cloud.githubusercontent.com/assets/23615360/22276459/144d20a2-e2f7-11e6-84cc-64a2c3e89781.png)
![-2](https://cloud.githubusercontent.com/assets/23615360/22276471/2f07328e-e2f7-11e6-8d06-c094611e7168.png)

When typed 테스트 on Openspades:
![-3](https://cloud.githubusercontent.com/assets/23615360/22276460/144e7312-e2f7-11e6-8e70-0e65f29d9237.png)
![-4](https://cloud.githubusercontent.com/assets/23615360/22276461/144fbf6a-e2f7-11e6-9e15-9a8414f50f71.png)

**Comment by feikname** [comment]
@kmsiapps Windows console doesn't officially support Unicode, it won't work out of the box.

You might try to use windows UTF-8 code page, but its not very stable. (Microsoft guys fault)

Type this before starting PySnip:
```batch
chcp 65001
```

There is also the possibility that `cmd.exe` is understanding what you type in OpenSpades, but just can't display it. You might solve this by changing its font to *Lucida console*.

**Comment by kmsiapps** [comment]
@feikname This error shows up :
> File "twisted\internet\base.pyo", line 793, in runUntilCurrent
          File "console.pyo", line 56, in get_input
        exceptions.LookupError: unknown encoding: cp65001

By the way, /log/log.txt shows the same result(saying ??? when typed Unicode chars).

**Comment by yvt** [comment]
To send/receive a Unicode text in a script, you need to implement [OpenSpades' protocol extension to send/receive Unicode strings](https://github.com/yvt/openspades/wiki/Network-Protocols#type-chatmessagestring). The recommended place to implement it is [`decode`](https://github.com/piqueserver/piqueserver/blob/e5c7b66ea66227aa7e60054750fa19636115641a/pyspades/common.pyx#L115) and [`encode`](https://github.com/piqueserver/piqueserver/blob/e5c7b66ea66227aa7e60054750fa19636115641a/pyspades/common.pyx#L111), but if you just want to send it anyway without modifying the server core, you can write a code like this (not recommended):

```python
# untested; let's hope this works
def encode_os_unicode_string(text): # type: (unicode) -> unicode
    # type: str
    utf8_encoded = text.encode('utf8')
    with_tag = '\xff' + utf8_encoded
    return with_tag.decode('cp437')
self.send_chat(encode_os_unicode_string(u"테스트"))
```

OpenSpades' implementation can be found [here](https://github.com/yvt/openspades/blob/2554eaad57144e7fdef048ed92c9edb9984e9ef1/Sources/Client/NetClient.cpp#L94-L114).

> You need to have Unicode enabled in OpenSpades besides having the pak, if I'm not mistaken.

It's enabled by default no matter whether font paks are installed or not. In fact, the built-in font has always included a small number of non-ascii glyphs (i.e., accented letters from Latin-1). OpenSpades just renders `□` if a character cannot be displayed using any installed font. 

> That's like user agent checking, which is generally from soft eng perspective is not that good of a route. A possibility to check for specific features would be better (doesSupportUnicode packet for example)

A future version of OpenSpades will support the enhanced variant of *Version Get* packet, which can be used to [query whether Unicode is supported](https://github.com/yvt/openspades/wiki/Network-Protocols#version-info-payload-2-client-feature-flags-1-since-01tbd). You'll need to resort to user agent checking for the presently available versions.

**Comment by kmsiapps** [comment]
Maybe we can translate server console messages after dropping support for py2?

**Comment by NotAFile** [comment]
I'm not sure how much sense it makes to translate console/logging messages. It just makes things harder to debug and then we will get shit like bug reports in korean.

**Comment by kmsiapps** [comment]
bug reports in korean, that makes sense lol
